### PR TITLE
XRT-2156 Fixes for data value comparison function

### PR DESCRIPTION
**Pull request by SteveOss**

### CODE REVIEW

**Comment by SteveOss** [comment]
@BinduRao2018 Still need to merge AzureSphere update

**Review by BinduRao2018 — COMMENTED** [review]
comments in-line

File: include/iot/data.h [hunk 1]
```diff
@@ -2013,13 +2015,14 @@ extern bool iot_data_equal_raw (const iot_data_t * data1, const iot_data_t * dat
 extern int iot_data_compare (const iot_data_t * data1, const iot_data_t * data2);
 
 /**
- * @brief Compare two data instances regardless of value datatype, returning whether the fist is less than, equal to or greater than the second
+ * @brief Compare two data instances, returning whether the fist is less than, equal to or greater than the second. Different integer
```
[the remark this answers is not in the document]
Comment (SteveOss) [reply include/iot/data.h:2018]: Fixed

File: include/iot/data.h [hunk 2]
```diff
@@ -2013,13 +2015,14 @@ extern bool iot_data_equal_raw (const iot_data_t * data1, const iot_data_t * dat
...
+ * or floating types are compared by value.
  *
  * @param  data1 Input data1 (can be NULL)
  * @param  data2 Input data2 (can be NULL)
  * @return       Returns zero if data1 equals data2, a value less than zero data1 less than data2, a value greater than zero if data1 greater than data2
```
[the remark this answers is not in the document]
Comment (SteveOss) [reply include/iot/data.h:2023]: 1 - Making it clear that data instances for some types (int and float) may be compared by value.
2 - Minus 2 is never returned function asserts if can't compare values. Should never reach this return, is there to keep compiler happy.

File: src/c/data.c [hunk 3]
```diff
@@ -503,69 +508,120 @@ extern const iot_data_t * iot_data_get_metadata (const iot_data_t * data, const
...
+    case IOT_DATA_INT32: uval.f64 = in.i32; break;
+    case IOT_DATA_UINT32: uval.f64 = in.ui32; break;
+    case IOT_DATA_INT64: uval.i64 = in.i64; break;
+    case IOT_DATA_UINT64: uval.ui64 = in.ui64; break;
+    case IOT_DATA_FLOAT32: uval.f64 = in.f32; break;
+    case IOT_DATA_FLOAT64: uval.f64 = in.f64; break;
+    default: uval.ui64 = in.bl ? 1u : 0u; break;
+  }
+  // Check value within max/min for target type
+  if ((out_type < IOT_DATA_FLOAT64) && (VALUE () > max_vals[out_type] || VALUE () < min_vals[out_type])) return false;
```
[the remark this answers is not in the document]
Comment (SteveOss) [reply src/c/data.c:541]: Don't have to check range for double as already holds the biggest and smallest values we can represent

File: include/iot/data.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (BinduRao2018) [reply include/iot/data.h:2023]: `if` missing in `a value less than zero data1 less than data2`

**Review by BinduRao2018 — APPROVED** [review]
LGTM, Hv left one unresolved comment

**Review by BinduRao2018 — APPROVED** [review]
LGTM

**Pull request merged**